### PR TITLE
metrics: fix error metrics and add instantaneous speed

### PR DESCRIPTION
**Pull request by lhy1024**
Signed-off-by: lhy1024 <admin@liudos.us>

### What problem does this PR solve?

Before this pr, `totalBytesRate` is not the total byte rate of all hot regions. This pr will fix it.

This pr also add `instantaneous` to get instantaneous byte rate with the heartbeat

image

The green one is `totalBytesRate`, which means the total byte rate of all hot regions.

The purple one is  `instantaneous`, which means that the instantaneous byte rate with the heartbeat.

The blue one is store bytes rate, which means that bytes rate with processed.

It also is a part of #3201

### Check List

Tests

- Unit test

### CODE REVIEW

**Comment by nolouch** [comment]
It's too many lines and hard to distinguish

File: server/statistics/hot_regions_stat.go [hunk 1]
```diff
@@ -15,8 +15,10 @@ package statistics
 
 // HotPeersStat records all hot regions statistics
 type HotPeersStat struct {
-	TotalBytesRate float64       `json:"total_flow_bytes"`
-	TotalKeysRate  float64       `json:"total_flow_keys"`
-	Count          int           `json:"regions_count"`
+	StoreBytesRate float64       `json:"store-bytes-rate"`
```
Comment (rleungx) [inline server/statistics/hot_regions_stat.go:18]: Why does hot peer stats need to have StoreBytesRate

Comment (lhy1024) [reply server/statistics/hot_regions_stat.go:18]: emmm, its name is confusing.

**Comment by lhy1024** [comment]
> It's too many lines and hard to distinguish

If no necessary, we will not show them.

File: server/cluster/coordinator.go [hunk 2]
```diff
@@ -453,6 +454,21 @@ func (c *coordinator) resetSchedulerMetrics() {
 	schedulerStatusGauge.Reset()
 }
 
+func collectStoreMetrics(rw, storeAddress, storeLabel string, storeID uint64, getStats func(storeID uint64) (string, *statistics.HotPeersStat)) {
+	typ, stat := getStats(storeID)
```
Comment (disksing) [inline server/cluster/coordinator.go:458]: why not pass in `typ` and `stat` directly?

File: server/statistics/store_collection.go [hunk 3]
```diff
@@ -108,14 +108,16 @@ func (s *storeStatistics) Observe(store *core.StoreInfo, stats *StoresStats) {
...
-
-	storeWriteRateByte, storeReadRateByte := storeFlowStats.GetBytesRate()
-	storeStatusGauge.WithLabelValues(storeAddress, id, "store_write_rate_bytes").Set(storeWriteRateByte)
-	storeStatusGauge.WithLabelValues(storeAddress, id, "store_read_rate_bytes").Set(storeReadRateByte)
-	storeWriteRateKey, storeReadRateKey := storeFlowStats.GetKeysWriteRate(), storeFlowStats.GetKeysReadRate()
-	storeStatusGauge.WithLabelValues(storeAddress, id, "store_write_rate_keys").Set(storeWriteRateKey)
-	storeStatusGauge.WithLabelValues(storeAddress, id, "store_read_rate_keys").Set(storeReadRateKey)
-
+	collect := func(getBytesRate, getKeysRate func() (float64, float64), tail string) {
+		storeWriteRateByte, storeReadRateByte := getBytesRate()
```
Comment (disksing) [inline server/statistics/store_collection.go:112]: same question.

File: server/statistics/store_hot_peers_infos.go [hunk 4]
```diff
@@ -21,3 +21,21 @@ type StoreHotPeersInfos struct {
...
 // StoreHotPeersStat is used to record the hot region statistics group by store.
 type StoreHotPeersStat map[uint64]*HotPeersStat
+
+// GetStoreStatAsPeer returns stat as peer from the corresponding store.
+func (info *StoreHotPeersInfos) GetStoreStatAsPeer(storeID uint64) (string, *HotPeersStat) {
+	stat, ok := info.AsPeer[storeID]
+	if ok {
+		return "as_peer", stat
+	}
+	return "as_peer", &HotPeersStat{}
```
Comment (disksing) [inline server/statistics/store_hot_peers_infos.go:31]: maybe this way is better?
```
if !ok {
  stat = &HotPeersStat{}
}
return "as_peer", stat
```

File: server/cluster/coordinator.go (hunk 2 above)
Comment (lhy1024) [reply server/cluster/coordinator.go:458]: because of so much line....

Comment (lhy1024) [reply server/cluster/coordinator.go:458]: Do you think we should revert them?

File: server/cluster/coordinator.go [hunk 5]
```diff
@@ -453,6 +454,21 @@ func (c *coordinator) resetSchedulerMetrics() {
 	schedulerStatusGauge.Reset()
 }
 
+func collectStoreMetrics(rw, storeAddress, storeLabel string, storeID uint64, getStats func(storeID uint64) (string, *statistics.HotPeersStat)) {
```
Comment (rleungx) [inline server/cluster/coordinator.go:457]: The function name does not quite match the functionality.

File: server/cluster/coordinator.go [hunk 6]
```diff
@@ -453,6 +454,21 @@ func (c *coordinator) resetSchedulerMetrics() {
...
+	typ, stat := getStats(storeID)
+	hotSpotStatusGauge.WithLabelValues(storeAddress, storeLabel, strings.Join([]string{"total", rw, "bytes", typ}, "_")).Set(stat.TotalBytesRate)
+	hotSpotStatusGauge.WithLabelValues(storeAddress, storeLabel, strings.Join([]string{"total", rw, "keys", typ}, "_")).Set(stat.TotalKeysRate)
+	hotSpotStatusGauge.WithLabelValues(storeAddress, storeLabel, strings.Join([]string{"hot", rw, "region", typ}, "_")).Set(float64(stat.Count))
+}
+
+func collectPendings(rw, storeAddress, storeLabel string, storeID uint64, pendings map[uint64]schedulers.Influence) {
+	infl := pendings[storeID]
+	// TODO: add to tidb-ansible after merging pending influence into operator influence.
```
Comment (rleungx) [inline server/cluster/coordinator.go:466]: Can we remove it now?

File: metrics/grafana/pd.json [hunk 7]
```diff
@@ -3629,7 +3629,7 @@
           "timeFrom": null,
           "timeRegions": [],
           "timeShift": null,
-          "title": "Total written bytes on hot peer Regions",
+          "title": "Total write bytes on hot peer Regions",
```
Comment (rleungx) [inline metrics/grafana/pd.json:3632]: Why?

**Review by disksing — APPROVED** [review]

**Comment by lhy1024** [comment]
@rleungx @disksing PTAL

**Review by rleungx — APPROVED** [review]

**Pull request merged**